language-plutus-core/src/Control/Monad/Trans/Inner.hs

```diff
@@ -0,0 +1,64 @@
+-- | Getting monad transormers out of 'Traversable' 'Monad's.
+
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+
+module Control.Monad.Trans.Inner
+    ( InnerT (..)
+    , forBind
+    , yield
+    , mapInnerT
+    ) where
+
+import           Control.Monad
+import           Control.Monad.Except
+
+forBind :: (Monad m, Traversable m, Applicative f) => m a -> (a -> f (m b)) -> f (m b)
+forBind a f = join <$> traverse f a
+
+-- | Turn a 'Traversable' 'Monad' into a monad transformer.
```

Comment:
I wonder if we could use `DerivingVia` to use this without having to actually have it around and do all the consequent wrapping and unwrapping?

Reply:
I'm also not wild on the name. What about something like `TraversableMonad`?

Reply:
There are a lot of tantalizing comments in this thread, but I can't find an actual library: https://www.reddit.com/r/haskell/comments/4nc19n/cross_section_of_the_monad_and_traversable_class/

Reply:
> Does something like this not exist on Hackage already?

It should exist, but I was unable to find anything.

> I wonder if we could use DerivingVia to use this without having to actually have it around and do all the consequent wrapping and unwrapping?

It's not a big deal to write a few lines of trivial type-driven code and `DerivingVia` is only supported by the very recent versions of GHC, so if we decide to make a library out of this, many people will not be able to use it.

> I'm also not wild on the name. What about something like `TraversableMonad`?

This sounds like a constraint rather than the name of a data type.

Reply:
> It's not a big deal to write a few lines of trivial type-driven code and DerivingVia is only supported by very recent versions of GHC, so if we decide to make a library out of this, many people will not be able to use it.

Realistically, we're going to support one version of GHC and that's going to be at least 8.6. I don't mind if you want to avoid `DerivingVia` for now, but this does seem like a prime use case.

> This sounds like a constraint rather than the name of a data type.

Well, the only reason this type exists is to guide the typeclass solver. I think there's some precedent for naming some things after the inference rule that they "enable".

Reply:
> Realistically, we're going to support one version of GHC and that's going to be at least 8.6. I don't mind if you want to avoid DerivingVia for now, but this does seem like a prime use case.

Yeah, I'm not against using `DerivingVia` in our code base at all and I'm not afraid of using it. I just think that what we have in this particular module can be uploaded to hackage eventually and *there* we don't want to have `DerivingVia` for compatibility reasons.

> Well, the only reason this type exists is to guide the typeclass solver. I think there's some precedent for naming some things after the inference rule that they "enable".

I'm not sure I've seen this. Do you have a reference? Also `TraversableMonad` sounds like a `Monad` that is `Traversable` rather than a `Monad` which you get from two `Monad`s, one of which is `Traversable`.

Reply:
Hm, I'm sure I saw this but maybe it was just in the endless stream of `DerivingVia` examples from @iceland_jack's twitter feed :sweat_smile: 

I agree `TraversableMonad` isn't great, but it at least conveys a bit more information than `InnerT`, which says absolutely nothing. Probably it should be something like `MonadComposeViaTraversable` or something, but that's unwieldy.